### PR TITLE
Do not allow to reuse deletion tokens (fixes #1171)

### LINKED ISSUE #1171: Reusing a token on paginated delete removes other records
I don't think we should allow reusing tokens on paginated delete operations, but if we do, I think we should just return the tombstones returned on the first time it was called.

```bash
# create three records
> http post localhost:8888/v1/buckets -a a:a
> http post localhost:8888/v1/buckets -a a:a
> http post localhost:8888/v1/buckets -a a:a
(...)

gsurita-30820:~ gsurita$ http delete localhost:8888/v1/buckets?_limit=1 -a a:a
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Backoff, Retry-After, Content-Length, Alert
Content-Length: 73
Content-Type: application/json
Date: Sun, 26 Mar 2017 13:04:04 GMT
Etag: "1490533444819"
Last-Modified: Sun, 26 Mar 2017 13:04:04 GMT
Next-Page: http://localhost:8888/v1/buckets?_limit=1&_token=eyJsYXN0X3JlY29yZCI6eyJsYXN0X21vZGlmaWVkIjoxNDkwNTMzNDQ0ODE5fSwib2Zmc2V0IjoxfQ%3D%3D
Server: waitress
Total-Records: 3
X-Content-Type-Options: nosniff

{
    "data": [
        {
            "deleted": true,
            "id": "wKQI-XKA",
            "last_modified": 1490533444819
        }
    ]
}

gsurita-30820:~ gsurita$ http delete "http://localhost:8888/v1/buckets?_limit=1&_token=eyJsYXN0X3JlY29yZCI6eyJsYXN0X21vZGlmaWVkIjoxNDkwNTMzNDQ0ODE5fSwib2Zmc2V0IjoxfQ%3D%3D" -a a:a
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Backoff, Retry-After, Content-Length, Alert
Content-Length: 73
Content-Type: application/json
Date: Sun, 26 Mar 2017 13:04:18 GMT
Etag: "1490533458436"
Last-Modified: Sun, 26 Mar 2017 13:04:18 GMT
Server: waitress
Total-Records: 2
X-Content-Type-Options: nosniff

{
    "data": [
        {
            "deleted": true,
            "id": "aM2SK729",
            "last_modified": 1490533458436
        }
    ]
}

# Same token of the last request (response is different)
gsurita-30820:~ gsurita$ http delete "http://localhost:8888/v1/buckets?_limit=1&_token=eyJsYXN0X3JlY29yZCI6eyJsYXN0X21vZGlmaWVkIjoxNDkwNTMzNDQ0ODE5fSwib2Zmc2V0IjoxfQ%3D%3D" -a a:a
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Backoff, Retry-After, Content-Length, Alert
Content-Length: 73
Content-Type: application/json
Date: Sun, 26 Mar 2017 13:04:19 GMT
Etag: "1490533459708"
Last-Modified: Sun, 26 Mar 2017 13:04:19 GMT
Server: waitress
Total-Records: 1
X-Content-Type-Options: nosniff

{
    "data": [
        {
            "deleted": true,
            "id": "ENUcJpjm",
            "last_modified": 1490533459708
        }
    ]
}
```

### PR DESCRIPTION
**Pull request by leplatrem**
Fixes #1171

### CODE REVIEW

**Comment by Natim** [comment]
I wish we could avoid querying the cache for that. I we use JWT token, we can add the timestamp and use this value to see if the token expired. Because of the signature nobody can forge a token.

**Comment by leplatrem** [comment]
But how do you prevent using it twice?

**Comment by Natim** [comment]
> But how do you prevent using it twice?

Why would you want such a thing?

**Comment by leplatrem** [comment]
> > But how do you prevent using it twice?
>
> Why would you want such a thing?

See original issue description.

File: kinto/core/resource/__init__.py [hunk 1]
```diff
@@ -1041,21 +1042,35 @@ def _extract_pagination_rules_from_token(self, limit, sorting):
...
                 error_msg = '_token has invalid content'
+
+            # We don't want pagination tokens to be reused several times (#1171).
+            # The cache backend is used to keep track of "nonces".
+            if self.request.method.lower() == "delete" and error_msg is None:
+                registry = self.request.registry
+                if registry.cache.get(nonce) is None:
+                    error_msg = '_token was already used or has expired.'
+                else:
+                    registry.cache.delete(nonce)
```
Comment (glasserc) [inline kinto/core/resource/__init__.py]: Does the `delete` after `get` open us up to race conditions?

Comment (leplatrem) [inline kinto/core/resource/__init__.py]: Yes, a pity that the delete method does not return anything...

Comment (glasserc) [inline kinto/core/resource/__init__.py]: It's still better than what we have and I'd be happy to merge it if there's a FIXME comment explaining the race and why we can't fix it.

**Comment by leplatrem** [comment]
Now requires #1232  for tests to pass

**Pull request merged**